optimism/src/ramlookup.rs

```diff
@@ -0,0 +1,119 @@
+use ark_ff::{Field, One, Zero};
+use kimchi_msm::{LookupTableID, MVLookup};
+
+/// Enum representing the two different modes of a RAMLookup
+#[derive(Copy, Clone, Debug)]
+pub enum LookupMode {
+    Read,
+    Write,
+}
+
+/// Struct containing a RAMLookup
+#[derive(Clone, Debug)]
+pub struct RAMLookup<T, ID: LookupTableID + Send + Sync + Copy> {
+    /// The table ID corresponding to this lookup
+    pub(crate) table_id: ID,
+    /// Whether it is a read or write lookup
+    pub(crate) mode: LookupMode,
+    /// The number of times that this lookup value should be added to / subtracted from the lookup accumulator.
+    pub(crate) magnitude: T,
+    /// The columns containing the content of this lookup
+    pub(crate) value: Vec<T>,
+}
+
+impl<T, ID> RAMLookup<T, ID>
+where
+    T: Clone
+        + std::ops::Add<T, Output = T>
+        + std::ops::Sub<T, Output = T>
+        + std::ops::Mul<T, Output = T>
+        + std::fmt::Debug
+        + One
+        + Zero,
+    ID: LookupTableID + Send + Sync + Copy,
+{
+    /// Creates a new RAMLookup from a mode, a table ID, a magnitude, and a value
+    pub fn new(mode: LookupMode, table_id: ID, magnitude: T, value: &[T]) -> Self {
+        Self {
+            mode,
+            table_id,
+            magnitude,
+            value: value.to_vec(),
+        }
+    }
+
+    /// Returns the numerator corresponding to this lookup in the MVLookup argument
+    pub fn numerator(&self) -> T {
+        match self.mode {
+            LookupMode::Read => T::zero() - self.magnitude.clone(),
```

Comment:
Doesn't `-self.magnitude.clone()` work?

Reply:
Unluckily not because that would mean that Self::Variable in the Env traits implements Neg, but that is not possible when Variable = u32 as it is the case in the witness of mips for example.